addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -46,3 +88,181 @@ func (r *CalicoConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		For(&cniv1alpha1.CalicoConfig{}).
 		Complete(r)
 }
+
+// ReconcileCalicoConfig reconciles CalicoConfig data values secrets
+func (r *CalicoConfigReconciler) ReconcileCalicoConfig(
+	ctx context.Context,
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig) (_ ctrl.Result, retErr error) {
+
+	var patchCRD bool
+
+	patchHelper, err := clusterapipatchutil.NewHelper(calicoConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// patch CalicoConfig before returning the function
+	defer func() {
+		// patchCRD will be true if finalizer or owner reference is added/removed
+		if patchCRD {
+			r.Log.Info("Patching CalicoConfig")
+			if err := patchHelper.Patch(ctx, calicoConfig.DeepCopy()); err != nil {
+				r.Log.Error(err, "Error patching CalicoConfig")
+				retErr = err
+			}
+		}
+
+		if err := r.Client.Status().Update(ctx, calicoConfig.DeepCopy()); err != nil {
+			r.Log.Error(err, "Error updating CalicoConfig Status")
+		}
+	}()
+
+	// if CalicoConfig is marked for deletion, then delete the data value secret
+	if !calicoConfig.GetDeletionTimestamp().IsZero() {
+		if err := r.ReconcileCalicoConfigDelete(ctx, calicoConfig, &patchCRD); err != nil {
+			r.Log.Error(err, "Error reconciling CalicoConfig delete to delete data value secret")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// reconcile CalicoConfig by creating the data value secret
+	if err := r.ReconcileCalicoConfigNormal(ctx, cluster, calicoConfig, &patchCRD); err != nil {
+		r.Log.Error(err, "Error reconciling CalicoConfig to create data value secret")
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("CalicoConfig successfully reconciled")
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCalicoConfigNormal reconciles CalicoConfig by creating data values secret
+func (r *CalicoConfigReconciler) ReconcileCalicoConfigNormal(
+	ctx context.Context,
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	patchCRD *bool) (retErr error) {
+
+	// add finalizer to CalicoConfig
+	*patchCRD = util.AddFinalizerToCRD(r.Log, constants.CalicoAddonName, calicoConfig)
+
+	// add owner reference to CalicoConfig if not already added by TanzuClusterBootstrap Controller
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               cluster.Kind,
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(true),
```

Comment:
Question: Setting this field to be true means we block workload cluster deletion until CalicoConfig resource gets deleted. That makes me whether it is better to NOT block a workload cluster from deleting, calicoconfig_controller could be smart to do the cleanup if the referenced cluster does not exist. We check the cluster existence in L71.

Reply:
Thanks for the comment. Removed this and added logic to remove orphaned CalicoConfig resource on line 75. Does that look good or garbage collector itself would take care of orphaned CalicoConfig resource deletion?

Reply:
Update: I was wrong before, the orphan CalicoConfig would be cleaned up automatically since it has ownerRef field set.